Changing the structure of the assessments page (#1030)
fairlearn/fairlearn: docs/user_guide/assessment.rst

Introduction
============

The goal of fairness assessment is to answer the question: Which groups of 
people may be disproportionately negatively impacted by an AI system and in 
what ways?

The steps of the assesment are as follows:

> 1. Identify harms

Reviewer: ```suggestion
1. Identify types of harms
```

2. Identify the groups that might be harmed
3. Quantify harms
4. Compare quantified harms across the groups

We next examine these four steps in more detail.

> 1. Identify harms

Reviewer: ```suggestion
1. Identify types of harms
```

-----------------

> 

Reviewer: We may want to add a reference to the types of harms in the previous user guide section 
https://fairlearn.org/main/user_guide/fairness_in_machine_learning.html#types-of-harms

Author: Done!

For example, in a system for screening job applications, qualified candidates 
that are automatically rejected experience an allocation harm. In a 

> speech-to-text transcription system, high error rates constitute harm in the 

Reviewer: ```suggestion
speech-to-text transcription system, disparities in word error rates for different groups may result in harms due to differences in the 
```

> quality of service.

Reviewer: Maybe add a caveat about how systems can lead to multiple harms at the same time. Feel free to re-use language or link to the SciPy tutorial overview slides!
https://github.com/fairlearn/talks/blob/main/2021_scipy_tutorial/overview.pdf

Author: I added a caveat at the end of this bullet!

2. Identify the groups that might be harmed
-------------------------------------------

In most applications, we consider demographic groups including historically 
marginalized groups (e.g., based on gender, race, ethnicity). We should also 

> consider groups that are relevant to a particular application. For example, for 

Reviewer: ```suggestion
consider groups that are relevant to a particular use case or deployment context. For example, for 
```

> speech-to-text transcription, groups based on the regional dialect or being a  

Reviewer: ```suggestion
speech-to-text transcription, this might include groups who speak a regional dialect or people who are a  
```

native or a non-native speaker.

It is also important to consider group intersections, for example, in addition
to considering groups according to gender and groups according to race, it is 
also important to consider their intersections (e.g., Black women, Latinx 
nonbinary people, etc.).

3. Quantify harms
-----------------

Define metrics that quantify harms or benefits:

> * In job screening scenario, we need to quantify the number of candidates that are classified as "negative" (not recommended for the job), but whose true label is "positive" (they are qualified). One possible metric is the false negative rate: fraction of qualified candidates that are screened out.

Reviewer: Maybe add something about first needing to understand the construct validity of what is considered "qualified" (can be a brief sentence with a link to the [construct validity section](https://fairlearn.org/main/user_guide/fairness_in_machine_learning.html#construct-validity) of the user guide)

Author: Added a brief sentence under bullet 3 that links to the Fairness in Machine learning page.

> * In speech-to-text scenario, the harm could be measured by word error rate, number of mistakes in a transcript divided by the overall number of words.

Reviewer: ```suggestion
* For a speech-to-text application, the harm could be measured by disparities in the word error rate for different group, measured by the number of mistakes in a transcript divided by the overall number of words.
```

> Note that in some cases, harms might not be readily identifiable. If you 

Reviewer: Is this issue you're raising here about how to _identify_ harms? Or whether some harms can be _quantified_? I ask because for representational harms, for instance, it is much easier to identify those harms (e.g., with stereotypes in language generation models) than it is to quantify those, which is an open area of research

Author: Maybe both? It might be difficult to identify which harms will appear downstream long after a system is implemented (like indirect effects). But I think the intent of this sentence is to discuss harms that are difficult to quantify, so I have changed the sentence accordingly.

> choose to use a proxy to represent the harm, check the proxy regularly 

Reviewer: So, we don't really explain proxies in our docs, other than briefly in the Boston housing dataset section and in the tutorial (which people reading this might not have read). 

Is this something we want to include here? If so, I think we should expand this to say more about what a proxy is in this case, and when and how people might want to identify proxies for sensitive features in their data (or not).

Author: I am fine with leaving it in or taking it out! I added a paragraph explaining what a proxy might look like in this case. If this isn't clear enough, I am fine with scrapping it. I've included the paragraph below, please let me know your thoughts.

> Note that in some cases, harms might not be readily identifiable. In these cases, we might choose to use a closely related variable to stand in for the missing variable. For example, suppose that in the job screening scenario, we have data on whether the candidate passes the first two stages, but not if they are ultimately recommended for the job. An alternative measure of harm might be if the candidate does not pass the first stage of the screen. If you choose to use a proxy variable to represent the harm, check the proxy variable regularly to ensure it remains useful over time.

Reviewer 2: Suggesting further edits to highlight the definition and role of a proxy in this setting:

```
Note that in some cases, harms might not be readily quantifiable. 
In these cases, we might choose to use a closely related variable,
often called a "proxy",  
to stand in for the missing variable. For example, suppose that in 
the job screening scenario, we have data on whether the candidate passes 
the first two stages, but not if they are ultimately recommended for the 
job.
As an alternative to the unobservable final recommendation, we could
therefore measure the harm using the proxy variable indicating whether
the candidate passes the first stage of the screen.
If you choose to use a proxy variable to 
represent the harm, check the proxy variable regularly to ensure it 
remains useful over time.
```

to ensure it remains useful over time.

4. Compare quantified harms across the groups
---------------------------------------------

The workhorse of fairness assessment are disaggregated metrics, which are 
metrics evaluated on slices of data. For example, to measure harms due to 
errors, we would begin by evaluating the errors on each slice of the data that 

> corresponds to a group. If some of the groups are seeing much larger errors 

Reviewer: I wonder if this is one place where we can raise the issue of what is considered "much larger" and the different approaches to that and their implications (e.g., the issue raised in the [paper](https://arxiv.org/abs/2202.09519) @LeJit mentioned on the call today). This might be a big ask for this PR, though, so we could maybe revisit this in future issues and PRs. @hildeweerts and @LeJit what do you think?

Reviewer 3: I think we could potentially add something that mentions the four-fifths rule in context of U.S. federal employment law, but adds a disclaimer that what constitutes "much larger" is highly context-dependent and a threshold that is applicable in one domain may not translate to a different application.

Author: In light of the conversation today, I'm thinking a path forward would be to table this addition until we have added more information about epistemic trespassing to the user guide. Then we could reference that page in this section. @mmadaio @LeJit thoughts?

Author: Keeping as is and resolving the comment - thanks!

different groups may result in harms due to differences in the quality of service.
Note that one system can lead to multiple harms, and different types of 
harms are not mutually exclusive. For more information, review 
Fairlearn's `2021 SciPy tutorial <https://github.com/fairlearn/talks/blob/main/2021_scipy_tutorial/overview.pdf>`_.

Identify the groups that might be harmed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In most applications, we consider demographic groups including historically 
marginalized groups (e.g., based on gender, race, ethnicity). We should also 
consider groups that are relevant to a particular use case or deployment context. For example, for 
speech-to-text transcription, this might include groups who speak a regional dialect or people who are a  
native or a non-native speaker.

It is also important to consider group intersections, for example, in addition
to considering groups according to gender and groups according to race, it is 
also important to consider their intersections (e.g., Black women, Latinx 
nonbinary people, etc.).

Quantify harms
^^^^^^^^^^^^^^

Define metrics that quantify harms or benefits:

> * In job screening scenario, we need to quantify the number of candidates that are classified as "negative" (not recommended for the job), but whose true label is "positive" (they are "qualified"). One possible metric is the false negative rate: fraction of qualified candidates that are screened out. Note that before we attempt to classify candidates, we need to determine the construct validity of the "qualified" status; more information on construct validity can be found in :ref:`construct_validity`.

Reviewer 4: ```suggestion
* In a job screening scenario, we need to quantify the number of candidates that are classified as "negative" (not recommended for the job), but whose true label is "positive" (they are "qualified"). One possible metric is the false negative rate: fraction of qualified candidates that are screened out. Note that before we attempt to classify candidates, we need to determine the construct validity of the "qualified" status; more information on construct validity can be found in :ref:`construct_validity`.
```

* For a speech-to-text application, the harm could be measured by disparities in the word error rate for different group, measured by the number of mistakes in a transcript divided by the overall number of words.

Note that in some cases, harms might not be readily quantifiable. 

> In these cases, we might choose to use a closely related variable,

Reviewer 4: It was not immediately clear to me that the quantifiability of the harm is directly related to a missing variable. An additional sentence to explain this might be useful (perhaps we can also relate it to some of the abstraction traps?).

Author: Ah, I understand. I've edited the sentences as follows. Updated text is italicized.

"Note that in some cases, harms might not be readily quantifiable. 
_Occasionally, another variable in our dataset provides a close
approximation to the phenomenon we seek to measure._
In these cases, we might choose to use that closely related variable,
often called a "proxy", to stand in for the missing variable. 
For example, suppose that in 
the job screening scenario, we have data on whether the candidate passes 
the first two stages, but not if they are ultimately recommended for the 
job. 

As an alternative to the unobservable final recommendation, we could
therefore measure the harm using the proxy variable indicating whether
the candidate passes the first stage of the screen.
If you choose to use a proxy variable to 
represent the harm, check the proxy variable regularly to ensure it 
remains useful over time. Our section on :ref:`construct_validity`
describes how to determine whether a  
proxy variable measures the intended construct in a meaningful 
and useful way. _Also be careful to ensure that the proxy measure 
you choose does not abstract away the social context  
of the problem you seek to solve. 
For more on abstraction traps, review :ref:`abstraction_traps`._"

Please let me know if this lines up with what you're thinking!

often called a "proxy",  
to stand in for the missing variable. For example, suppose that in 
the job screening scenario, we have data on whether the candidate passes 
the first two stages, but not if they are ultimately recommended for the 
job.

As an alternative to the unobservable final recommendation, we could
therefore measure the harm using the proxy variable indicating whether
the candidate passes the first stage of the screen.
If you choose to use a proxy variable to 
represent the harm, check the proxy variable regularly to ensure it 
remains useful over time. Our section on :ref:`construct_validity`
describes how to determine whether a  
proxy variable measures the intended construct in a meaningful 
and useful way.

Compare quantified harms across the groups
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The workhorse of fairness assessment are disaggregated metrics, which are 

Reviewer 4: ```suggestion
The workhorse of fairness assessment in Fairlearn are disaggregated metrics, which are 
```

metrics evaluated on slices of data. For example, to measure harms due to 
errors, we would begin by evaluating the errors on each slice of the data that 
corresponds to a group. If some of the groups are seeing much larger errors 
than other groups, we would flag this as a fairness harm.

To summarize the disparities in errors (or other metrics), we may want to 
report quantities such as the difference or ratio of the metric values between 
the best and the worst slice. In settings where the goal is to guarantee 
certain minimum quality of service across all groups (such as speech recognition), 
it is also meaningful to report the worst performance across all considered groups.

> For example, when comparing false negative rate across groups defined by race, 

Reviewer 4: ```suggestion
For example, when comparing the false negative rate across groups defined by race, 
```

Identify types of harms
^^^^^^^^^^^^^^^^^^^^^^^

See :ref:`types_of_harms` for a guide to types of fairness-related harms. 
For example, in a system for screening job applications, qualified candidates 
that are automatically rejected experience an allocation harm. In a 
speech-to-text transcription system, disparities in word error rates for 
different groups may result in harms due to differences in the quality of service.
Note that one system can lead to multiple harms, and different types of 
harms are not mutually exclusive. For more information, review 
Fairlearn's `2021 SciPy tutorial <https://github.com/fairlearn/talks/blob/main/2021_scipy_tutorial/overview.pdf>`_.

Identify the groups that might be harmed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In most applications, we consider demographic groups including historically 
marginalized groups (e.g., based on gender, race, ethnicity). We should also 
consider groups that are relevant to a particular use case or deployment context. For example, for 
speech-to-text transcription, this might include groups who speak a regional dialect or people who are a  
native or a non-native speaker.

> It is also important to consider group intersections, for example, in addition

Reviewer 5: I'd love to see a reference to Kimberle Crenshaw's work on intersectionality here. Wdyt?

Author: Sure! I added a sentence _Kimberlé Crenshaw's `seminal work on intersectionality <https://www.jstor.org/stable/1229039>`_ 
offers a thorough background on this topic._, which links to the JSTOR file, in lieu of creating a reference at the bottom of the page. I don't see this paper in the refs.bib file though, and I think we should add it there because it's referenced in multiple places!

Reviewer 4: We only recently moved to BibTex which would explain the absence of the reference. I think it would be great if you could add it as part of this PR! You could also possibly adjust the references on the other places in the user guide you've spotted - but we can also leave that to another PR.

Author: Ah, gotcha. I'll update the .bib file with the Crenshaw paper and add it to this PR. I do think there's room for a larger task of adding all the referenced links throughout the user guide to that file. I think this is a good first issue, so I can open up that PR for another contributor to work on!

Reviewer 4: Agreed! Great suggestion to open an issue for this!

* For a speech-to-text application, the harm could be measured by disparities in the word error rate for different group, measured by the number of mistakes in a transcript divided by the overall number of words.

Note that in some cases, the outcome we seek to measure is not 
directly available. 
Occasionally, another variable in our dataset provides a close 
approximation to the phenomenon we seek to measure. 
In these cases, we might choose to use that closely related variable, 
often called a "proxy", to stand in for the missing variable. 
For example, suppose that in the job screening scenario, 
we have data on whether the candidate passes the first two stages, 
but not if they are ultimately recommended for the job. 

As an alternative to the unobservable final recommendation, we could
therefore measure the harm using the proxy variable indicating whether
the candidate passes the first stage of the screen.
If you choose to use a proxy variable to 
represent the harm, check the proxy variable regularly to ensure it 

> remains useful over time. Our section on :ref:`construct_validity`

Reviewer 5: Love it! Goodhart's law! https://en.wikipedia.org/wiki/Goodhart's_law?wprov=sfti1

describes how to determine whether a  
proxy variable measures the intended construct in a meaningful 
and useful way. It is important to ensure that the proxy is suitable 
for the social context of the problem you seek to solve. 

> In particular, be careful of falling into one of the :refabstraction_traps. 

Reviewer 5: ```suggestion
In particular, be careful of falling into one of the :ref:abstraction_traps. 
```
And back quotes also I think (?)

Compare quantified harms across the groups
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The workhorse of fairness assessment in Fairlearn are disaggregated metrics, which are 

Reviewer 5: Perhaps it's me (non-native speaker here!) but workhorse doesn't sound right to me.
Key piece? Center piece?

Author: Changed to centerpiece!

metrics evaluated on slices of data. For example, to measure harms due to 
errors, we would begin by evaluating the errors on each slice of the data that 
corresponds to a group. If some of the groups are seeing much larger errors 
than other groups, we would flag this as a fairness harm.

To summarize the disparities in errors (or other metrics), we may want to 
report quantities such as the difference or ratio of the metric values between 
the best and the worst slice. In settings where the goal is to guarantee 
certain minimum quality of service across all groups (such as speech recognition), 
it is also meaningful to report the worst performance across all considered groups.

For example, when comparing the false negative rate across groups defined by race, 

> we may summarize our findings with a table like the following:

Reviewer 5: Having worked with these for a while now I don't trust any such tables anymore because they don't tell me the size of each group. Perhaps we should have an example that has that, too?

Author: I can add some hypothetical group population counts, if that would work?

Author: I added the population counts as well as the warning "Note that the these statistics must 
be drawn from a large enough sample size to draw meaningful conclusions.".

* For a speech-to-text application, the harm could be measured by disparities in the word error rate for different group, measured by the number of mistakes in a transcript divided by the overall number of words.

Note that in some cases, the outcome we seek to measure is not 
directly available. 
Occasionally, another variable in our dataset provides a close 
approximation to the phenomenon we seek to measure. 
In these cases, we might choose to use that closely related variable, 
often called a "proxy", to stand in for the missing variable. 
For example, suppose that in the job screening scenario, 
we have data on whether the candidate passes the first two stages, 
but not if they are ultimately recommended for the job. 

> As an alternative to the unobservable final recommendation, we could

Reviewer 4: Nitpick: is the final recommendation unobservable or just unobserved?

Author: I think unobserved, good catch! I just changed it.

that are automatically rejected experience an allocation harm. In a 
speech-to-text transcription system, disparities in word error rates for 
different groups may result in harms due to differences in the quality of service.
Note that one system can lead to multiple harms, and different types of 
harms are not mutually exclusive. For more information, review 
Fairlearn's `2021 SciPy tutorial <https://github.com/fairlearn/talks/blob/main/2021_scipy_tutorial/overview.pdf>`_.

Identify the groups that might be harmed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In most applications, we consider demographic groups including historically 
marginalized groups (e.g., based on gender, race, ethnicity). We should also 
consider groups that are relevant to a particular use case or deployment context. For example, for 
speech-to-text transcription, this might include groups who speak a regional dialect or people who are a  
native or a non-native speaker.

It is also important to consider group intersections, for example, in addition
to considering groups according to gender and groups according to race, it is 
also important to consider their intersections (e.g., Black women, Latinx 

> nonbinary people, etc.). Kimberlé Crenshaw's `work on intersectionality <https://www.jstor.org/stable/1229039>`_ 

Reviewer 4: I believe this is the only edit we still require. Could you add a screenshot of the final page @alliesaizan?

```suggestion
nonbinary people, etc.). Kimberlé Crenshaw's work on intersectionality :footcite:p:`crenshaw1991intersectionality`
```

Author: The final HTML page? I am having some issues building the documentation on my end so that may be tricky.

Reviewer 5: Any eay we can help eith thr issues?

Author: I'll try to troubleshoot it further this weekend. I primarily work on a Windows machine but can also try it on a Mac. and see if the HTML pages render. When I run the Windows [commands on the website](https://fairlearn.org/v0.7.0/contributor_guide/contributing_documentation.html), under docs>_build> the html folder is created, but its only content is a .doctrees folder. I'm guessing the issue is with sphinx itself, but I can give more details after I have looked into it a bit more.

Reviewer 5: I am also using Windows. I'd try deleting the entire build directory so that it can start afresh.

Author: I was able to figure it out. I updated my Python version and was able to build the docs (some of the plots weren't rendering in my old Python environment, which was holding up the build.) Images of the first few sections are below.

![image](https://user-images.githubusercontent.com/15962391/162581534-30c810b5-3383-4df4-8923-f2aae811d598.png)

![image](https://user-images.githubusercontent.com/15962391/162581553-4e4ad48a-d899-48a8-8407-c62a804bef52.png)

![image](https://user-images.githubusercontent.com/15962391/162581571-149293bb-b464-4935-abcd-01511465737b.png)

Assessment
==========

> 

Reviewer 6: The document begins with two headers: Assessment and Introduction. Perhaps we should hide the Assessment header, or is this a bad idea?

Author: I don't think we need it, so I've removed it. I can revert though if needed.

Introduction
------------

The goal of fairness assessment is to answer the question: Which groups of 
people may be disproportionately negatively impacted by an AI system and in 
what ways?

> The steps of the assesment are as follows:

Reviewer 6: The steps of the introduction of the assessment, right?

Author: Oh no, this is straight from the tutorial. I think this is supposed to be the steps to conducting an assessment in general

1. Identify harms
2. Identify the groups that might be harmed
3. Quantify harms
4. Compare quantified harms across the groups

We next examine these four steps in more detail.

1. Identify harms
^^^^^^^^^^^^^^^^^

For example, in a system for screening job applications, qualified candidates 
that are automatically rejected experience an allocation harm. In a 
speech-to-text transcription system, high error rates constitute harm in the 
quality of service.

2. Identify the groups that might be harmed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In most applications, we consider demographic groups including historically 
marginalized groups (e.g., based on gender, race, ethnicity). We should also 
consider groups that are relevant to a particular application. For example, for 
speech-to-text transcription, groups based on the regional dialect or being a  
native or a non-native speaker.

It is also important to consider group intersections, for example, in addition
to considering groups according to gender and groups according to race, it is 
also important to consider their intersections (e.g., Black women, Latinx 
nonbinary people, etc.).

> 3. Quantify harms

Reviewer 7: Should we add a note to this section that some harms might not be readily quantifiable. And also that if using a proxy for the harm, users will need to monitor whether the proxy remains useful over time?

Author: I added a few sentences to this effect in item 3.

Define metrics that quantify harms or benefits:

* In job screening scenario, we need to quantify the number of candidates that are classified as "negative" (not recommended for the job), but whose true label is "positive" (they are qualified). One possible metric is the false negative rate: fraction of qualified candidates that are screened out.

* In speech-to-text scenario, the harm could be measured by word error rate, number of mistakes in a transcript divided by the overall number of words.

4. Compare quantified harms across the groups
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The workhorse of fairness assessment are disaggregated metrics, which are 
metrics evaluated on slices of data. For example, to measure harms due to 
errors, we would begin by evaluating the errors on each slice of the data that 
corresponds to a group. If some of the groups are seeing much larger errors 
than other groups, we would flag this as a fairness harm.

To summarize the disparities in errors (or other metrics), we may want to 
report quantities such as the difference or ratio of the metric values between 
the best and the worst slice. In settings where the goal is to guarantee 

> certain minimum quality of service (such as speech recognition), it is also 

Reviewer 7: Slight nitpick: if there's a minimal QoS goal, then surely 'minimum across all groups' _is_ the key fairness metric?

Author: Added "across all groups" to this sentence (so now reads "minimum quality of service _across all groups_"), please let me know if this is sufficient

meaningful to report the worst performance across all considered groups.

For example, when comparing false negative rate across groups defined by race, 
we may summarize our findings with a table like the following:

.. list-table::
   :header-rows: 1
   :widths: 7 30
   :stub-columns: 1

   *  - 
      - false negative rate (FNR)

   *  - AfricanAmerican
      - 0.43

   *  - Caucasian
      - 0.44

   *  - Other
      - 0.52

   *  - Unknown
      - 0.67

   *  - largest difference
      - 0.24 (best is 0.0)

   *  - smallest ratio
      - 0.64 (best is 1.0)

   *  - maximum (worst-case) FNR
      - 0.67

> Ungrouped metrics

Reviewer 7: Why not 'Disaggregated' here?

* For a speech-to-text application, the harm could be measured by disparities in the word error rate for different group, measured by the number of mistakes in a transcript divided by the overall number of words.

Note that in some cases, harms might not be readily quantifiable. 
In these cases, we might choose to use a closely related variable,
often called a "proxy",  
to stand in for the missing variable. For example, suppose that in 
the job screening scenario, we have data on whether the candidate passes 
the first two stages, but not if they are ultimately recommended for the 
job.

As an alternative to the unobservable final recommendation, we could
therefore measure the harm using the proxy variable indicating whether
the candidate passes the first stage of the screen.
If you choose to use a proxy variable to 
represent the harm, check the proxy variable regularly to ensure it 

> remains useful over time.

Reviewer 3: Optional, but we might want to link to our section of the User Guide about [construct validity](https://fairlearn.org/v0.7.0/user_guide/fairness_in_machine_learning.html#what-is-construct-validity) here since it's a framework for assessing the "usefulness" of a proxy variable.

Author: Done!

speech-to-text transcription system, disparities in word error rates for 
different groups may result in harms due to differences in the quality of service.
Note that one system can lead to multiple harms, and different types of 
harms are not mutually exclusive. For more information, review 
Fairlearn's `2021 SciPy tutorial <https://github.com/fairlearn/talks/blob/main/2021_scipy_tutorial/overview.pdf>`_.

2. Identify the groups that might be harmed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In most applications, we consider demographic groups including historically 
marginalized groups (e.g., based on gender, race, ethnicity). We should also 
consider groups that are relevant to a particular use case or deployment context. For example, for 
speech-to-text transcription, this might include groups who speak a regional dialect or people who are a  
native or a non-native speaker.

It is also important to consider group intersections, for example, in addition
to considering groups according to gender and groups according to race, it is 
also important to consider their intersections (e.g., Black women, Latinx 
nonbinary people, etc.).

> 3. Quantify harms

Reviewer 5: The numbers come automatically.

Author: I removed the numbering and added the `.. sectnum::` parameter; is that what you mean?

.. _metrics_with_grouping:

Metrics with grouping using :code:`MetricFrame`
-----------------------------------------------

In a typical fairness assessment, each row of input data will have an associated
group label :math:`g \in G`, and we will want to know how the metric behaves
for each :math:`g`. To help with this, Fairlearn provides a class that takes
an existing (ungrouped) metric function and applies it to each group within a
set of data.

This data structure, :class:`fairlearn.metrics.MetricFrame`, enables evaluation 
of disaggregated metrics. In its simplest form :class:`fairlearn.metrics.MetricFrame` 
takes four arguments:
* metric_function with signature metric_function(y_true, y_pred)
* y_true: array of labels
* y_pred: array of predictions

> * sensitive_features: array of sensitive feature values

Reviewer 6: Is this intended to be rendered as a list?

Author: Yes, I fixed the spacing issues. Thanks!

.. _metrics_with_grouping:

> Metrics with grouping using :code:`MetricFrame`

Author: Ok, I have made the update to "Disaggregated metrics using :code:MetricFrame". Sorry about that!

-----------------------------------------------

In a typical fairness assessment, each row of input data will have an associated
group label :math:`g \in G`, and we will want to know how the metric behaves
for each :math:`g`. To help with this, Fairlearn provides a class that takes

> an existing (ungrouped) metric function and applies it to each group within a

Reviewer 5: Perhaps mention "like scikit-learn's metrics"?

Author: Added a reference to :module:`sklearn.metrics`, but let me know if that's not the correct domain directive.

Author: @hildeweerts I have addressed both your comments! I added the function references and replaced "ungrouped" with "disaggregated" to more directly link that sentence to the section above.

set of data.

This data structure, :class:`fairlearn.metrics.MetricFrame`, enables evaluation 
of disaggregated metrics. In its simplest form :class:`fairlearn.metrics.MetricFrame` 
takes four arguments:

> * metric_function with signature metric_function(y_true, y_pred)

Reviewer 5: Can we put these inside :code:?

Author: Done!

    >>> print("recall by groups = ", grouped_metric.by_group.to_dict())
    recall by groups =  {'a': 0.0, 'b': 0.5, 'c': 0.75, 'd': 0.0}

The disaggregated metrics are stored in a pandas Series 
:code:`grouped_metric.by_group`. Note that the overall recall is the same 

> as that calculated above in the Ungrouped Metric section, while the 'by group' 

Reviewer 4: ```suggestion
as that calculated above in the Ungrouped Metric section, while the 'by_group'
```

    >>> print("recall by groups = ", grouped_metric.by_group.to_dict())
    recall by groups =  {'a': 0.0, 'b': 0.5, 'c': 0.75, 'd': 0.0}

> The disaggregated metrics are stored in a pandas Series 

Reviewer 5: I'd suggest using intersphinx for pandas.Series using :class:

Author: Done!

    >>> print("ratio in recall = ", grouped_metric.ratio(method='between_groups'))    
    ratio in recall =  0.0

> Multiple metrics in one :code:`MetricFrame`

Reviewer 4: ```suggestion
Multiple metrics in a single :code:`MetricFrame`
```

-------------------------------------------

A single instance of :class:`fairlearn.metrics.MetricFrame` can evaluate multiple
metrics simultaneously by providing the `metrics` argument with a 
dictionary of desired metrics. The disaggregated metrics are then stored in a 

> pandas DataFrame. Note that :func:`fairlearn.metrics.count` can 

Reviewer 4: ```suggestion
pandas DataFrame. Note that :class:`pandas.DataFrame` can 
```

    d    0...
    Name: metric, dtype: float64

Multiclass metrics
^^^^^^^^^^^^^^^^^^

We may also be interested in multiclass classification. However, typical group
fairness metrics such as equalized odds and demographic parity are only defined
for binary classification. One way to measure fairness in the multiclass
scenario is to define one-to-one or one-to-rest classifications for each group
and calculate the metrics on this instead. Alternatively, we can use predefined
metrics for multiclass classification. For example, accuracy is a multiclass
metric that we can use through scikit-learn's :py:func:`sklearn.metrics.accuracy_score`
in combination with a :code:`MetricFrame` as follows:

.. doctest:: assessment_metrics

    >>> from sklearn.metrics import accuracy_score
    >>> from fairlearn.metrics import MetricFrame
    >>> y_mult_true = [0,1,2,1,3,0,1,3,0,2,1,2,0,0,1,3]
    >>> y_mult_pred = [0,1,1,2,3,0,1,0,0,2,1,2,3,0,0,2]
    >>> mf = MetricFrame(metric=accuracy_score,
    ...                  y_true=y_mult_true, y_pred=y_mult_pred,
    ...                  sensitive_features=group_membership_data)

>     >>> print(mf.by_group) # series with accuracy for each sensitive group

Reviewer 4: The [doc tests failed](https://app.circleci.com/pipelines/github/fairlearn/fairlearn/2692/workflows/5fc3e423-7c46-4e84-9944-25a17fba4ea1/jobs/5148) for this line, because the expected output format is floats:

```
Expected:
    sensitive_feature_0
    a         1.0
    b         0.5
    c    0.428571
    d         1.0
    Name: accuracy_score, dtype: object
Got:
    sensitive_feature_0
    a    1.000000
    b    0.500000
    c    0.428571
    d    1.000000
    Name: accuracy_score, dtype: float64
```

Reviewer 4: Awesome, let me know if anything is unclear @alliesaizan!

Author: Looks like it passed all the checks! 🎉
